presto-geospatial-toolkit/src/test/java/com/facebook/presto/geospatial/TestKDBTree.java

```diff
@@ -0,0 +1,226 @@
+        }
+
+        KDBTree tree = buildKDBTree(25, 100, extent, rectangles.build());
+
+        Map<Integer, Rectangle> leafNodes = tree.findAllLeafNodes();
+        assertEquals(leafNodes.size(), 2);
+        assertEquals(leafNodes.keySet(), ImmutableSet.of(0, 1));
+        assertEquals(leafNodes.get(0), new Rectangle(0, 0, 4, 4.5));
+        assertEquals(leafNodes.get(1), new Rectangle(0, 4.5, 4, 9));
+
+        // points inside and outside partitions
+        assertPartitions(tree, new Rectangle(1, 1, 1, 1), ImmutableSet.of(0));
+        assertPartitions(tree, new Rectangle(1, 6, 1, 6), ImmutableSet.of(1));
+        assertPartitions(tree, new Rectangle(5, 1, 5, 1), ImmutableSet.of());
+
+        // point on the border separating two partitions
+        assertPartitions(tree, new Rectangle(1, 4.5, 1, 4.5), ImmutableSet.of(0, 1));
+
+        // rectangles
+        assertPartitions(tree, new Rectangle(1, 1, 2, 2), ImmutableSet.of(0));
+        assertPartitions(tree, new Rectangle(1, 6, 2, 7), ImmutableSet.of(1));
+        assertPartitions(tree, new Rectangle(1, 1, 2, 5), ImmutableSet.of(0, 1));
+        assertPartitions(tree, new Rectangle(5, 1, 6, 2), ImmutableSet.of());
+    }
+
+    private void assertPartitions(KDBTree kdbTree, Rectangle envelope, Set<Integer> partitions)
+    {
+        Map<Integer, Rectangle> matchingNodes = kdbTree.findIntersectingLeafNodes(envelope);
+        assertEquals(matchingNodes.size(), partitions.size());
+        assertEquals(matchingNodes.keySet(), partitions);
+    }
+
+    @Test
+    public void testEvenDistribution()
+    {
+        testEvenDistribution(0, 0);
+        testEvenDistribution(1, 2);
+    }
+
+    private void testEvenDistribution(double width, double height)
+    {
+        Rectangle extent = new Rectangle(0, 0, 9, 4);
+        ImmutableList.Builder<Rectangle> rectangles = ImmutableList.builder();
+        for (int x = 0; x < 10; x++) {
+            for (int y = 0; y < 5; y++) {
+                rectangles.add(new Rectangle(x, y, x + width, y + height));
+            }
+        }
+
+        KDBTree tree = buildKDBTree(10, 100, extent, rectangles.build());
+
+        Map<Integer, Rectangle> leafNodes = tree.findAllLeafNodes();
+        assertEquals(leafNodes.size(), 6);
+        assertEquals(leafNodes.keySet(), ImmutableSet.of(0, 1, 2, 3, 4, 5));
+        assertEquals(leafNodes.get(0), new Rectangle(0, 0, 2.5, 2.5));
+        assertEquals(leafNodes.get(1), new Rectangle(0, 2.5, 2.5, 4));
+        assertEquals(leafNodes.get(2), new Rectangle(2.5, 0, 4.5, 4));
+        assertEquals(leafNodes.get(3), new Rectangle(4.5, 0, 7.5, 2.5));
+        assertEquals(leafNodes.get(4), new Rectangle(4.5, 2.5, 7.5, 4));
+        assertEquals(leafNodes.get(5), new Rectangle(7.5, 0, 9, 4));
+    }
+
+    @Test
+    public void testSkewedDistribution()
+    {
+        testSkewedDistribution(0, 0);
+        testSkewedDistribution(1, 2);
+    }
+
+    private void testSkewedDistribution(double width, double height)
```

Comment:
Make sure this part is tested

> // Too many objects are crowded together. Can't split.

Reply:
@arhimondr Added more tests to cover this code path.